ssl/statem/statem_lib.c

```diff
@@ -1007,6 +1007,28 @@ int ssl_choose_client_version(SSL *s, int version)
 }
 
 /*
+ * ssl_verify_client_session_version - Verify that provided SSL_SESSION matches
+ * the protocol negotiated with server.
+ *
+ * @s: client SSL handle.
+ *
+ * Returns 1 when it matches, 0 otherwise
+ */
+int ssl_verify_client_session_version(SSL *s)
+{
+  /* Empty session and no EAP-FAST */
+  if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {
```

Comment:
Oh. This might actually be hiding another bug. Sorry, my bad! I should have realized this when I pointed out the `session->ssl_version = version` line. Because `ssl_get_new_session` is called *before* version resumption in `tls_construct_client_hello`, the `ssl_version` field isn't initialized right.

One option is to remove this check and restore the ssl_version assignment, but in line 1132, where you know the session is fresh. Another option, and what I did for BoringSSL, is to take the `ssl_get_new_session` call out of `tls_construct_client_hello` and uncondition the other one. But that's a more involved change. (It means `s->session` can be NULL in more states.)

I think you'd be able to reproduce this with tests that ensure session resumption happens at all versions. https://rt.openssl.org/Ticket/Display.html?id=4392 is an example of what goes wrong. (OpenSSL currently has a bug where `ssl_version` isn't initialized right in DTLS on the server half.)

(Actually, you might want to play with https://mta.openssl.org/pipermail/openssl-dev/2016-March/005779.html. The bug you're fixing here is caught by BoringSSL tests.)

NB: I also have no clue about EAP-FAST and how this interacts. We've removed that code because it was crazy. :-)